### PR TITLE
Store the current module hash in a system table

### PR DESCRIPTION
**Pull request by kim**
# Description of Changes

Adds a new system table `st_module` which stores the hash of the stdb module currently associated with the database. The hash is updated _transactionally_ when the database is initialized or updated (that is, the table does not have any rows until the `init` procedure ran successfully).

`st_module` also associates an "epoch" number with the row, which is used to store a _fencing token_. As sketched out in the changes to standalone, this allows to use an external (distributed) locking service to safely order lifecycle operations on the database. We can thus guarantee _at-most-once_ semantics for `init` and `update` reducers.

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API
 - [X] This is a breaking change to the database internals

Because a new system table is introduced, compatibility with existing databases cannot be maintained. The patch is against `cloud-next`, which contains breaking changes anyways.

### CODE REVIEW

**Comment by mamcx** [comment]
I recommend you wait for my [PR](https://github.com/clockworklabs/SpacetimeDB/pull/267) that changes and simplifies the use of system tables.

**Comment by kim** [comment]
I’d love to, but this one is also fairly critical for cloud to work correctly. If you haven’t considered already, I would also greatly appreciate if new system tables could be introduced after the db was bootstrapped (I.e. a gap in the reserved sequence range).

File: crates/core/src/db/datastore/locking_tx_datastore/mod.rs [hunk 1]
```diff
@@ -1673,6 +1675,37 @@ impl Locking {
...
+        match tx.lock.iter(&ST_MODULE_ID)?.next() {
+            None => Ok(None),
+            Some(data) => {
+                let row = StModuleRow::try_from(data.view())?;
+                Ok(Some(row.program_hash))
+            }
+        }
+    }
+
+    pub(crate) fn set_module_hash(&self, tx: &mut MutTxId, token: u64, hash: Hash) -> Result<(), DBError> {
```
Comment (kim) [inline crates/core/src/db/datastore/locking_tx_datastore/mod.rs:1689]: These should probably not be inherents, but I couldn't make up my mind yet how to make them fit into the trait structure.

File: crates/core/src/db/datastore/traits.rs [hunk 2]
```diff
@@ -498,3 +498,32 @@ pub trait MutTxDatastore: TxDatastore + MutTx {
         row: ProductValue,
     ) -> Result<ProductValue>;
 }
+
+/// Describes a programmable [`TxDatastore`].
+///
+/// A programmable datastore is one which has a program of some kind associated
+/// with it.
+pub trait Programmable: TxDatastore {
```
Comment (kim) [inline crates/core/src/db/datastore/traits.rs:506]: Naming?

File: crates/core/src/db/datastore/traits.rs [hunk 3]
```diff
@@ -498,3 +498,32 @@ pub trait MutTxDatastore: TxDatastore + MutTx {
...
+    /// `set_module_hash` with respect to a distributed locking service.
+    type FencingToken: Eq + Ord;
+
+    /// Update the [`Hash`] of the program currently associated with the
+    /// datastore.
+    ///
+    /// The operation runs within the transactional context `tx`. The fencing
+    /// token `fence` must be verified to be greater than in any previous
+    /// invocations of this method.
+    fn set_program_hash(&self, tx: &mut Self::MutTxId, fence: Self::FencingToken, hash: Hash) -> Result<()>;
```
Comment (gefjon) [inline crates/core/src/db/datastore/traits.rs:528]: Does this cause the program itself to be updated, or only the stored hash?

File: crates/standalone/src/lib.rs [hunk 4]
```diff
@@ -487,87 +481,150 @@ impl StandaloneEnv {
...
+            Err(e) => {
+                stdb.rollback_tx(tx);
+                Err(e.into())
+            }
+
+            Ok(maybe_hash) => {
+                // Release tx due to locking semantics and acquire a control db
+                // lock instead.
+                stdb.commit_tx(tx)?;
+                let lock = self.lock_database_instance_for_update(instance.id)?;
```
Comment (gefjon) [inline crates/standalone/src/lib.rs:560]: This feels like a race condition...

**Review by gefjon — APPROVED** [review]

File: crates/core/src/db/datastore/traits.rs (hunk 3 above)
Comment (kim) [reply crates/core/src/db/datastore/traits.rs:528]: Only the hash — but inside the transaction which updates the module.

File: crates/standalone/src/lib.rs (hunk 4 above)
Comment (kim) [reply crates/standalone/src/lib.rs:560]: It is, but we can’t thread through the tx. If a later thread is faster acquiring the lock, this update will not have any effect — that is, we guarantee at most once, not exactly once.

Comment (gefjon) [reply crates/standalone/src/lib.rs:560]: Ah, ok. Not ideal, but probably fine.

File: crates/core/src/db/datastore/traits.rs (hunk 3 above)
Comment (kim) [reply crates/core/src/db/datastore/traits.rs:551]: It's a good point, actually -- the datastore doesn't store its own module. That does make sense in a distributed setting, but is another piece to worry about for keeping a replica consistent. Food for thought...

**Pull request merged**